File: clients/python/pgque/consumer.py
# Copyright 2026 Nikolay Samokhvalov. Apache-2.0 license.
# PgQue includes code derived from PgQ (ISC license,
# Marko Kreen / Skype Technologies OU).

"""Consumer -- event-driven message consumer with LISTEN/NOTIFY support."""

import logging
import select
import signal
import threading
import time
from typing import Callable, Literal, Optional

import psycopg
from psycopg import sql

from .client import PgqueClient
from .types import Message

logger = logging.getLogger("pgque")

# Maximum time the LISTEN wait blocks before re-checking the stop flag.
# Bounds shutdown latency to roughly this many seconds.
_WAIT_SLICE_SECONDS = 0.5
# PostgreSQL int4 max; request the whole batch by default.
_DEFAULT_MAX_MESSAGES = 2_147_483_647


class Consumer:
    """Synchronous polling consumer with LISTEN/NOTIFY wakeup.

    Usage::

        consumer = Consumer(
            dsn="postgresql://localhost/mydb",
            queue="orders",
            name="order_processor",
        )

        @consumer.on("order.created")
        def handle_order(msg: Message):
            process_order(msg.payload)

        consumer.start()  # blocks until SIGTERM/SIGINT

    Handler return semantics:
        - If the handler returns without exception, the message is
          considered processed.
        - If the handler raises an exception, the message is nacked
          with the default retry_after.
        - If no handler is registered for a message type (and no
          default ``"*"`` handler exists), the message is nacked
          (sent to retry_queue, or to the dead-letter queue once
          ``queue_max_retries`` is exhausted). To ack unknown types
          instead, pass ``unknown_handler_policy="ack"``.

    After all messages in a batch have been dispatched, the batch is
    acked automatically.
    """

    def __init__(
        self,
        dsn: str,
        *,
        queue: str,
        name: str,
        poll_interval: int = 30,
        max_messages: int = _DEFAULT_MAX_MESSAGES,
        retry_after: int = 60,
        unknown_handler_policy: Literal["nack", "ack"] = "nack",
        subconsumer: Optional[str] = None,
        dead_interval: Optional[str] = None,
    ):
        self.dsn = dsn
        self.queue = queue
        self.name = name
        self.poll_interval = poll_interval
        self.max_messages = max_messages
        self.retry_after = retry_after
        if unknown_handler_policy not in ("nack", "ack"):
            raise ValueError(
                "unknown_handler_policy must be 'nack' or 'ack', "
                f"got {unknown_handler_policy!r}"
            )
        self._unknown_handler_policy = unknown_handler_policy

        # Experimental cooperative-consumers mode. When ``subconsumer`` is
        # set, the poll loop calls ``client.receive_coop(...)`` instead of
        # the normal ``receive(...)``. ``dead_interval`` is meaningless
        # outside coop mode and signals a programming error if provided.
        if dead_interval is not None and subconsumer is None:
            raise ValueError(
                "dead_interval is only valid in cooperative mode "
                "(set subconsumer=...)"
            )
        self.subconsumer = subconsumer
        self.dead_interval = dead_interval

        self._handlers: dict[str, Callable] = {}
        self._default_handler: Optional[Callable] = None
        self._running = False
        self._log = logging.getLogger(f"pgque.consumer.{name}")

    def on(self, event_type: str):
        """Decorator to register a handler for a given event type.

        Args:
            event_type: The ``pgque.message.type`` value to match.
                Use ``"*"`` to register a default/catch-all handler.
        """

        def decorator(func: Callable):
            if event_type == "*":
                self._default_handler = func
            else:
                self._handlers[event_type] = func
            return func

        return decorator

    def start(self) -> None:
        """Run the consume loop (blocks until SIGTERM/SIGINT).

        Opens its own connection, subscribes to LISTEN, and polls for
        batches. Each batch is processed and acked in a single
        transaction.
        """
        self._running = True

        # Graceful shutdown on signals; only main-thread invocations can
        # install signal handlers. When the consumer is run from a worker
        # thread (tests, embedded use), skip registration -- callers stop
        # via Consumer.stop().
        in_main_thread = threading.current_thread() is threading.main_thread()
        original_sigterm = None
        original_sigint = None

        def _stop(signum, frame):
            logger.info("received signal %s, shutting down", signum)
            self._running = False

        if in_main_thread:
            original_sigterm = signal.getsignal(signal.SIGTERM)
            original_sigint = signal.getsignal(signal.SIGINT)
            signal.signal(signal.SIGTERM, _stop)
            signal.signal(signal.SIGINT, _stop)

        try:
            with psycopg.connect(self.dsn, autocommit=True) as conn:
                # Subscribe for wakeup notifications
                channel = f"pgque_{self.queue}"
                conn.execute(sql.SQL("LISTEN {}").format(sql.Identifier(channel)))
                logger.info(
                    "consumer %s listening on %s (poll=%ds)",
                    self.name,
                    self.queue,
                    self.poll_interval,
                )

                while self._running:
                    processed = self._poll_once(conn)

                    if not self._running:
                        break

                    if processed:
                        # Non-empty batch: more batches may already be
                        # ticked (backlog). Their notifies fired while
                        # we were not listening, so waiting for a new
                        # NOTIFY would drain the backlog at one batch
                        # per poll_interval. Re-poll immediately until
                        # the queue comes back empty.
                        continue

                    # Wait for NOTIFY or poll_interval timeout in short
                    # bounded slices. psycopg's conn.notifies() can
                    # block uninterruptibly for the full timeout, which
                    # makes stop() slow and can miss prompt wakeups.
                    # Polling the underlying socket with
                    # select() lets us re-check _running every SLICE
                    # seconds and drain any pending NOTIFY immediately.
                    self._wait_for_notify_or_stop(conn)

        finally:
            if in_main_thread:
                signal.signal(signal.SIGTERM, original_sigterm)
                signal.signal(signal.SIGINT, original_sigint)

        logger.info("consumer %s stopped", self.name)

    def stop(self) -> None:
        """Request graceful shutdown (safe to call from another thread)."""
        self._running = False

    def _wait_for_notify_or_stop(self, conn: psycopg.Connection) -> None:
        """Wait up to ``poll_interval`` for a NOTIFY, in short slices.

        Returns early on any of:
          * a NOTIFY arrives (drained from the connection),
          * ``stop()`` flips ``_running`` to False,
          * ``poll_interval`` elapses cumulatively.

        Each slice is at most ``_WAIT_SLICE_SECONDS`` so ``stop()`` is
        observed within ~SLICE seconds of the call.
        """
        # Drain any NOTIFY already buffered in libpq from the prior
        # _poll_once (e.g. delivered alongside query results). Without
        # this, a buffered notify sits in libpq until the socket
        # becomes readable for some other reason -- select() won't see
        # it, and wakeup latency stretches out. Restores the implicit
        # entry-drain semantics of the old conn.notifies(timeout=...).
        drained = False
        for _notify in conn.notifies(timeout=0):
            drained = True
        if drained:
            return

        deadline = time.monotonic() + self.poll_interval
        fd = conn.fileno()
        while self._running:
            remaining = deadline - time.monotonic()
            if remaining <= 0:
                return
            slice_timeout = min(_WAIT_SLICE_SECONDS, remaining)
            # select() returns when the socket is readable (notify
            # delivered by the server) or when slice_timeout expires.
            # It is a thin wrapper around the OS poll, so it is cheap
            # and interruptible.
            r, _w, _x = select.select([fd], [], [], slice_timeout)
            if not self._running:
                return
            if r:
                # Drain pending notifications without blocking. A
                # zero timeout makes notifies() return immediately
                # after consuming whatever is buffered.
                for _notify in conn.notifies(timeout=0):
                    pass
                return

    def _poll_once(self, conn: psycopg.Connection) -> bool:
        """Receive one batch and dispatch messages.

        Returns True when a non-empty batch was processed and acked
        (the caller should re-poll immediately to drain any backlog),
        False when the queue was empty or the batch was left unacked
        for redelivery (the caller should wait for NOTIFY/timeout).

        If any per-message ``nack()`` raises, all remaining messages in
        the batch are still dispatched (their handlers run), but the
        batch is NOT acked at the end -- the receive transaction commits
        without finishing the batch, so PgQ redelivers the whole batch
        on the next poll. Without this guard, swallowing a nack failure
        and then acking would advance past the batch and silently drop
        the failed message.
        """
        # Use a transaction block for receive + ack
        with conn.transaction():
            client = PgqueClient(conn)
            if self.subconsumer is not None:
                msgs = client.receive_coop(
                    self.queue,
                    self.name,
                    self.subconsumer,
                    max_messages=self.max_messages,
                    dead_interval=self.dead_interval,
                )
            else:
                msgs = client.receive(
                    self.queue, self.name, self.max_messages
                )

            if not msgs:
                return False

            batch_id = msgs[0].batch_id
            logger.debug(
                "batch %d: %d message(s)", batch_id, len(msgs)
            )

            nack_failed = False

            for msg in msgs:
                handler = self._handlers.get(msg.type, self._default_handler)
                if handler is None:
                    if self._unknown_handler_policy == "ack":
                        self._log.warning(
                            "no handler for event type=%s ev_id=%s; acking",
                            msg.type,
                            msg.msg_id,
                        )
                        continue
                    self._log.warning(
                        "no handler for event type=%s ev_id=%s; nacking",
                        msg.type,
                        msg.msg_id,
                    )
                    try:
                        client.nack(
                            batch_id,
                            msg,
                            retry_after=self.retry_after,
                            reason=f"no handler for type={msg.type}",
                        )
                    except Exception:
                        nack_failed = True
                        self._log.exception(
                            "nack failed for unhandled msg_id=%d; "
                            "skipping batch ack so PgQ redelivers",
                            msg.msg_id,
                        )
                        continue
                    continue

                try:
                    handler(msg)
                except Exception:
                    self._log.exception(
                        "handler failed for msg_id=%d, nacking",
                        msg.msg_id,
                    )
                    try:
                        client.nack(
                            batch_id, msg, retry_after=self.retry_after
                        )
                    except Exception:
                        nack_failed = True
                        self._log.exception(
                            "nack failed for msg_id=%d; "
                            "skipping batch ack so PgQ redelivers",
                            msg.msg_id,
                        )
                        continue

            if nack_failed:
                # Do NOT ack -- redeliver on next poll. Report False so
                # the caller waits before re-polling instead of
                # redelivering the same batch in a hot loop.
                return False

            # pgque.ack returns 1 on success, 0 if the batch was already
            # finished or not found (stale/double ack, cross-consumer
            # race). Mirror the TS+Go consumers and log warn on 0; do
            # not treat it as an error.
            if client.ack(batch_id) == 0:
                logger.warning(
                    "pgque: ack batch %d returned 0 -- stale or double ack "
                    "(batch already finished or not found)",
                    batch_id,
                )

            return True

File: clients/ruby/lib/pgque/consumer.rb
# Copyright 2026 Nikolay Samokhvalov. Apache-2.0 license.
# PgQue includes code derived from PgQ (ISC license,
# Marko Kreen / Skype Technologies OU).

require "logger"

module Pgque
  class Consumer
    DEFAULT_MAX_MESSAGES = 2_147_483_647
    WAIT_SLICE_SECONDS = 0.5

    attr_reader :dsn, :queue, :name, :poll_interval, :max_messages,
                :retry_after, :subconsumer, :dead_interval

    attr_accessor :logger

    def initialize(dsn, queue:, name:, poll_interval: 30,
                   max_messages: DEFAULT_MAX_MESSAGES, retry_after: 60,
                   unknown_handler_policy: "nack", subconsumer: nil,
                   dead_interval: nil, logger: nil)
      @dsn = dsn
      @queue = queue
      @name = name
      @poll_interval = poll_interval
      @max_messages = max_messages
      @retry_after = retry_after

      unless ["nack", "ack"].include?(unknown_handler_policy.to_s)
        raise ArgumentError,
              "unknown_handler_policy must be 'nack' or 'ack', " \
              "got #{unknown_handler_policy.inspect}"
      end
      @unknown_handler_policy = unknown_handler_policy.to_s

      if dead_interval && subconsumer.nil?
        raise ArgumentError,
              "dead_interval is only valid in cooperative mode " \
              "(set subconsumer:)"
      end
      @subconsumer = subconsumer
      @dead_interval = dead_interval

      @handlers = {}
      @default_handler = nil
      # @running is a plain boolean. Ruby integer/boolean assignment
      # is atomic, and the only cross-thread interactions are the
      # signal trap and Consumer#stop flipping it false while the
      # main loop polls running? -- no ordering dependencies, so a
      # mutex would be overkill (and unsafe to enter from a signal
      # trap, which raises ThreadError on Mutex#synchronize).
      @running = false
      @stop_signum = nil
      @logger = logger || default_logger
    end

    def on(event_type, &block)
      raise ArgumentError, "block required for Consumer#on" unless block

      if event_type == "*"
        @default_handler = block
      else
        @handlers[event_type] = block
      end
      block
    end

    def start
      @running = true
      @stop_signum = nil

      in_main_thread = (Thread.current == Thread.main)
      original_handlers = {}

      # Signal traps run in a restricted context: Mutex#synchronize,
      # Logger#info, and most blocking code raise ThreadError. Keep
      # this proc to plain instance-variable writes; the main loop
      # logs the signal number after waking up.
      stop_proc = ->(signum) {
        @stop_signum = signum
        @running = false
      }

      if in_main_thread
        ["TERM", "INT"].each do |sig|
          original_handlers[sig] = Signal.trap(sig) { stop_proc.call(sig) }
        end
      end

      begin
        conn = PG.connect(@dsn)
        begin
          channel = "pgque_#{@queue}"
          conn.exec("LISTEN #{conn.escape_identifier(channel)}")
          @logger.info(
            "consumer #{@name} listening on #{@queue} (poll=#{@poll_interval}s)"
          )

          while running?
            poll_once(conn)
            break unless running?
            wait_for_notify_or_stop(conn)
          end

          if @stop_signum
            @logger.info("received signal #{@stop_signum}, shutting down")
          end
        ensure
          conn.close unless conn.finished?
        end
      ensure
        if in_main_thread
          original_handlers.each { |sig, h| Signal.trap(sig, h || "DEFAULT") }
        end
        @logger.info("consumer #{@name} stopped")
      end
    end

    def stop
      @running = false
    end

    def running?
      @running
    end

    # Public for testability; not part of the stable API.
    def poll_once(conn)
      conn.transaction do
        client = Client.new(conn)
        msgs =
          if @subconsumer
            client.receive_coop(
              @queue, @name, @subconsumer,
              max_messages: @max_messages,
              dead_interval: @dead_interval,
            )
          else
            client.receive(@queue, @name, @max_messages)
          end

        next if msgs.empty?

        batch_id = msgs[0].batch_id
        @logger.debug("batch #{batch_id}: #{msgs.size} message(s)")

        nack_failed = dispatch_batch(client, batch_id, msgs)

        next if nack_failed

        rowcount = client.ack(batch_id)
        if rowcount == 0
          @logger.warn(
            "pgque: ack batch #{batch_id} returned 0 -- stale or " \
            "double ack (batch already finished or not found)",
          )
        end
      end
    end

    private

    def dispatch_batch(client, batch_id, msgs)
      nack_failed = false
      msgs.each do |msg|
        handler = @handlers[msg.type] || @default_handler

        if handler.nil?
          if @unknown_handler_policy == "ack"
            @logger.warn(
              "no handler for event type=#{msg.type} ev_id=#{msg.msg_id}; " \
              "acking",
            )
            next
          end
          @logger.warn(
            "no handler for event type=#{msg.type} ev_id=#{msg.msg_id}; " \
            "nacking",
          )
          begin
            client.nack(batch_id, msg, retry_after: @retry_after,
                        reason: "no handler for type=#{msg.type}")
          rescue StandardError => e
            nack_failed = true
            @logger.error(
              "nack failed for unhandled msg_id=#{msg.msg_id}: " \
              "#{e.class}: #{e.message}",
            )
          end
          next
        end

        begin
          handler.call(msg)
        rescue StandardError => e
          @logger.error(
            "handler failed for msg_id=#{msg.msg_id}: " \
            "#{e.class}: #{e.message}",
          )
          begin
            client.nack(batch_id, msg, retry_after: @retry_after)
          rescue StandardError => e2
            nack_failed = true
            @logger.error(
              "nack failed for msg_id=#{msg.msg_id}: " \
              "#{e2.class}: #{e2.message}",
            )
          end
        end
      end
      nack_failed
    end

    def wait_for_notify_or_stop(conn)
      drained = false
      while conn.notifies
        drained = true
      end
      return if drained

      deadline = monotonic + @poll_interval
      while running?
        remaining = deadline - monotonic
        return if remaining <= 0

        slice = [WAIT_SLICE_SECONDS, remaining].min
        notification = conn.wait_for_notify(slice)
        return unless running?

        if notification
          while conn.notifies
            # drain any queued notifications
          end
          return
        end
      end
    end

    def monotonic
      Process.clock_gettime(Process::CLOCK_MONOTONIC)
    end

    # The default logger is effectively silent: it targets $stderr (so
    # messages never collide with application stdout) and ships at level
    # FATAL, which the consumer never emits. Set PGQUE_LOG_LEVEL=warn (or
    # info, debug, error) to see warnings/info from the consumer, or
    # pass logger: Logger.new(...) to Consumer.new for full control.
    def default_logger
      log = Logger.new($stderr)
      log.progname = "pgque.consumer.#{@name}"
      log.level = env_log_level || Logger::FATAL
      log
    end

    def env_log_level
      raw = ENV["PGQUE_LOG_LEVEL"]
      return nil if raw.nil?

      normalized = raw.strip.upcase
      return nil if normalized.empty?

      Logger.const_get(normalized)
    rescue NameError
      nil
    end
  end
end
